Allow = in config key names

nickva opened the pull request:
They are allowed only for the "k = v" format and not the "k=v" format. The idea is to split on " = " first, and if that fails to produce a valid kv pair we split on "=" as before.

To implement it, simplify the parsing logic and remove the undocumented multi-line config value feature. The continuation syntax is not documented anywhere and not used by our default.ini or in documentation.

Fix: #3319

jaydoane commented:
Just a couple nits: First, the title of the PR "Allow = in config key values" makes it sound like we're allowing the `=` character in the _values_, but I think you mean "key names"?

> There are allowed only for the `k = v` form and not `k=v`. 

Do you mean "They are allowed ..."? 

> The idea is to split on `=` first, and if that fails to produce a valid kv pair we split on `=` as before.

Should the first one contain spaces on either side of the equal sign, like ` = `?

> Also added 100% test coverage to the parser.

💯 awesome

jaydoane commented on src/config/src/config.erl:
```diff
@@ -414,89 +415,73 @@ is_sensitive(Section, Key) ->
+    % Empty section don't parse any lines until we're in a section
+    {Section, KVs};
+parse_fold(Line, {Section, KVs}) ->
+    case string:split(Line, " = ") of
```
Should we add documentation that describes this algorithm?

nickva replied:
Good idea!

jaydoane reviewed (approved):
The parsing code is much easier to understand now, although I think the way it first tries to parse the line first with ` = ` and then falls back to `=` is a bit subtle, and should be clearly documented.

nickva commented:
> Just a couple nits: First, the title of the PR "Allow = in config key values" makes it sound like we're allowing the `=` character in the _values_, but I think you mean "key names"?
> 

Good catch. Yeah that is ambiguous, I meant "key names".

> > There are allowed only for the `k = v` form and not `k=v`.
> 
> Do you mean "They are allowed ..."?
>

Yup! Silly mistake. Thanks for noticing!

> > The idea is to split on `=` first, and if that fails to produce a valid kv pair we split on `=` as before.
> 
> Should the first one contain spaces on either side of the equal sign, like `=`?
> 

I had it in the commit comment but then tried to make prettier in the PR comment messed it up instead. I'll return back to using quotes.

nickva commented:
Updated and added docs for it. 

When adding docs a typo (`[jwt_key]` should be `[jwt_keys]`) so fixed that along the way as well.

In addition manually tested that it's possible to set and get the value via the `_config/` API:

```
% curl -XPUT -H'Content-Type:application/json'  $DB/_node/_local/_config/s1/k1 -d '"v1"'
% curl -XPUT -H'Content-Type:application/json'  $DB/_node/_local/_config/s1/k=1 -d '"v1"'
% curl -XPUT -H'Content-Type:application/json'  $DB/_node/_local/_config/s1/k=== -d '"v1"'

% curl -s $DB/_node/_local/_config/s1 | jq '.'
{
  "k1": "v1",
  "k=1": "v1",
  "k===": "v1"
}

% cat dev/lib/node1/etc/local.ini
...
[s1]
k1 = v1
k=1 = v1
k=== = v1
```
